### PR TITLE
Selecting a new system selects the first mission

### PR DESCRIPTION
**Pull request by alextd**
## Summary
Selecting a new system on the mission map selects the first available mission

## Save File
This save file can be used to see the UI behavior:
[John Oliver.txt](https://github.com/endless-sky/endless-sky/files/8815926/John.Oliver.txt)

Load in, open mission map. Click Sol. See 6th mission 'tourists to Luna' selected. Click Diphda. See its mission selected. Click back to Sol. See the same mission for Luna selected. Before it would continue down past Diphda's mission, onto your accepted Sol mission instead of the new, available mission (and then you'd have to click through all your accepted missions to get back to available missions, or click the first mission and then back to Sol - ugh! Now, it just selects the one you want immediately)

-----------------------
**Feature:** This PR implements the feature request detailed and discussed in issue #4464 (and first attempted in PR #4414)

## Feature Details
Selecting a new system selects the first mission - the first available mission, or first accepted mission when none available

Before, it would continue from wherever your selected mission happens to be, and seemingly randomly select an accepted job instead of an available job. This was simply because the new available mission was above the currently selected mission, and it searches down only, and then finds an accepted mission before looking at that available mission. If a different mission were already selected, maybe that new available mission would be chosen. So now it's consistent - selecting a new system always selects the first mission.

Selecting an empty system doesn't count as a new system - it's effectively a non-click for which mission gets selected.

## UI Screenshots
image

## Usage Examples
As in the save file, when you have missions accepted and want to select new available missions - selecting the system goes to an available mission instead of seemingly randomly selecting your already-accepted mission.

## Testing Done
Clicked around a lot of systems. Clicked empty systems. Clicked looping over all missions at a system. Accepted missions in different orders. The first mission is consistently selected when selecting a new system, no matter where the previous mission was.

## Performance Impact
N/A.

### CODE REVIEW

**Review by Hurleveur — COMMENTED** [review]
I quite like the idea, it makes sense, however I think it should select the first acceptable mission instead of the absolute first one; like it currently selects the first available one when you open the map panel and directly click on a system with acceptable and unacceptable missions.

**Comment by alextd** [comment]
> it currently selects the first available one when you open the map panel and directly click on a system

I mean, no, this whole point of this PR is that the one it selected was unintuitive and seemingly random. There's definitely no check that the mission is unacceptable (greyed-out) or not. The fact you think it does, goes to show that it's brokenly unintuitive and this PR is sorely needed.

Tested: Load up this save, hit 'j', select Fah Soom ( bottom left )
[Halaster Blackcloack~BADSELECTED.txt](https://github.com/endless-sky/endless-sky/files/9051659/Halaster.Blackcloack.BADSELECTED.txt)
image

As for the actual feature of selecting an acceptable mission --- see my other PR #6915 for sorting the mission list, which also separates out unacceptable missions so the first mission in the list *is* an acceptable one.

**Review by RisingLeaf — APPROVED** [review]
Looks fine structure wise

**Review by Hurleveur — APPROVED** [review]
Agreed this is good to merge for me

File: source/MissionPanel.cpp [hunk 1]
```diff
@@ -406,7 +406,33 @@ bool MissionPanel::Click(int x, int y, int clicks)
...
 		}
-		while(options--)
+
+		// When clicking a new system, Select the first available mission
+		// (instead of continuing from wherever the iterator happens to be)
+		if((availableIt != available.end() && !Involves(*availableIt, system)) ||
+			(acceptedIt != accepted.end() && !Involves(*acceptedIt, system)))
+		{
+			auto firstExistingIt = find_if(available.begin(), available.end(),
+				[&system](const Mission &m) { return Involves(m, system);});
```
Comment (warp-core) [inline source/MissionPanel.cpp:416]:
```suggestion
				[&system](const Mission &m) { return Involves(m, system); });
```

File: source/MissionPanel.cpp [hunk 2]
```diff
@@ -406,7 +406,33 @@ bool MissionPanel::Click(int x, int y, int clicks)
...
+
+			if(firstExistingIt != available.end())
+			{
+				availableIt = firstExistingIt;
+				acceptedIt = accepted.end();
+			}
+			else
+			{
+				firstExistingIt = find_if(accepted.begin(), accepted.end(),
+					[&system](const Mission &m) { return m.IsVisible() && Involves(m, system);});
```
Comment (warp-core) [inline source/MissionPanel.cpp:426]:
```suggestion
					[&system](const Mission &m) { return m.IsVisible() && Involves(m, system); });
```

**Review by warp-core — COMMENTED** [review]
Style

File: source/MissionPanel.cpp [hunk 3]
```diff
@@ -460,7 +460,33 @@ bool MissionPanel::Click(int x, int y, int clicks)
 			else
 				acceptedIt = accepted.begin();
 		}
-		while(options--)
+
+		// When clicking a new system, Select the first available mission
```
Comment (Amazinite) [inline source/MissionPanel.cpp:464]:
```suggestion
		// When clicking a new system, select the first available mission
```

**Pull request merged**